zilliqa/src/consensus.rs

```diff
@@ -472,87 +548,111 @@ impl Consensus {
+            while let Some(mut txn) = txs.pop() {
+                txn.retries += 1;
+                temp_vec.push(txn.clone());
+
+                // if we have reached the block limit, break
+                if ret.len() >= self.config.consensus.block_tx_limit {
+                    break;
+                }
 
-            // If it is higher, mark it for a retry
-            if tx.transaction.nonce > nonce {
-                let retries = self.new_transactions_waiting.entry(tx.hash()).or_insert(0);
-                let _ = retries.add(1);
-                warn!(
-                        "Transaction nonce for tx {} is too high: {} when acct nonce is {}, tries: {}/{}",
-                        tx.hash(),
-                        tx.transaction.nonce,
-                        self.state.must_get_account(tx.from_addr).nonce,
-                        retries,
+                // if it is less than the iter nonce, we can skip it
+                // this just means there are 'priority txs' with the same nonce but different fees
+                if txn.nonce < iter_nonce {
+                    trace!("lower priority tx ignored {} vs {}", txn.nonce, iter_nonce);
+                    continue;
+                }
+
+                // If the nonce is too high, we will take another pass
+                if txn.nonce > iter_nonce {
+                    trace!(
+                        "TX nonce {} too high, iter nonce {}, retry {} of {}",
+                        txn.nonce,
+                        iter_nonce,
+                        txn.retries,
                         self.config.tx_retries
                     );
 
-                if *retries >= self.config.tx_retries {
-                    warn!(?tx, "Tranaction has exceeded retries and will been removed");
-                    tx_hashes_invalid.push(tx.hash());
-                    continue;
+                    if txn.retries >= self.config.tx_retries {
+                        warn!(?txn, "Tranaction has exceeded retries and all pending from this account will been removed");
+                        remove_all_txs = true;
+                    }
+
+                    // all other TXs will have a higher nonce, so we can break
+                    break;
                 }
-            }
 
-            // Tx nonce is good to add, we need to check that there isn't already a tx with the same
-            // from address, though (no need to drop though)
-            if tx_from_addrs.contains(&tx.from_addr) {
-                warn!(
-                    "Transaction from address {} already exists in block and will not be added",
-                    tx.from_addr
-                );
-                continue;
+                // tx nonce == iter nonce, so we can add it
+                match self.new_transactions.get(&txn.hash) {
+                    Some(tx) => {
+                        trace!("Adding tx to execute: {:?}", tx);
+                        ret.push(tx.clone());
```

Comment:
Note that by now, if I'm reading this correctly, we will have already pushed `txn` into `temp_vec` (L587). Thus at the same time as we're adding `tx` to the ret vec and executing it this block, we're also putting the hash back into the priority queue below (L638). This doesn't seem right - but let me know if I've misunderstood how this works.

Reply:
Hi, that was my intention. Hashes are removed later when the block is executed 'for real' and not just during construction. 

In my mind, it is better to keep the heap as-is for the reap, because you are not guaranteed your block will be executed after you construct it (you can time out or other stuff). If this happens, the TXs dissapear essentially.